src/main/scala/io/dronekit/cloud/SignRequestForAWS.scala

```diff
@@ -0,0 +1,302 @@
+  def createUri(uri: Uri): Uri = {
+    Uri(uri.scheme, uri.authority,
+      Path(uri.path.toString().replace("+","%20")),
+      Query(uri.query.toString().replace("+","%20").split('&').sorted.mkString("&")),
+      uri.fragment)
+  }
+
+  //SHA hashes the payload
+  def signPayload(payload: String): String = {
+    val md = MessageDigest.getInstance("SHA-256")
+    md.update(payload.getBytes("UTF-8"))
+    Hex.encodeHexString(md.digest())
+  }
+
+  /**
+   * Generate a canonical request string as defined in
+   * http://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html
+   * @param httpRequest The request to generate the string from
+   * @return A canonical string representation of the request to be used for signing
+   */
+  def createCanonicalRequest(httpRequest: HttpRequest): Future[String] = {
+    val contentsFuture =
+      if (httpRequest.entity.isKnownEmpty()) Future.successful("")
+      else httpRequest.entity.dataBytes.map(_.utf8String).runWith(Sink.head)
+    contentsFuture.map { entity =>
+      s"""${httpRequest.method.name}
+         |${httpRequest.uri.path.toString().replace("+", "%20")}
+         |${generateValidUriQuery(httpRequest.uri.query)}
+         |${generateCanonicalHeaders(httpRequest.headers :+ RawHeader("host", httpRequest.uri.authority.host.toString()))}
+         |${signPayload(entity)}""".stripMargin
+    }
+  }
+  //TODO: need to uri encode depending on what is being passed in
+  // replaces spaces with %20 and sorts the query parameters
+  def generateValidUriQuery(query: Query): String = {
+    s"""${query}""".replace("+","%20").split('&').sorted
+      .mkString("&")
+  }
+
+
+  // creates the string to sign based on aws protocol
+  def createStringToSign(httpRequest: HttpRequest, canonicalRequest: String, region: String, service: String): String = {
+    s"""${"AWS4-HMAC-SHA256"}
+       |${getCanonicalFormDate(httpRequest.headers)}
+       |${getCredentialScope(httpRequest.headers, httpRequest.uri, region, service)}
+       |${hashCanonicalRequest(canonicalRequest)}""".stripMargin
+  }
+
+  // get canonical date with time stamp
+  def getCanonicalFormDate(headers: Seq[HttpHeader]): String = {
+    val dateString = headers
+      .filter(_.lowercaseName() == "x-amz-date")
+      .map{header => header.lowercaseName() -> header.value().trim}
+      .map{case (name, values) => s"$values"}
+      .mkString("\n")
+    // how to choose something else if dateString is empty
+    if (dateString.length <= 1 ) {
+      getUTCTime()
+    } else {
+      dateString
+    }
+  }
+
+  //runs same algorithm as sign payload but on the canonicalRequest
+  def hashCanonicalRequest(canonicalRequest: String): String = {
+    val md = MessageDigest.getInstance("SHA-256")
+    md.update(canonicalRequest.getBytes("UTF-8"))
+    Hex.encodeHexString(md.digest())
+  }
+
+  //finds the relevant fields in the http request and returns the credential scope
```

Comment:
getCredentialScope and createSignatureKey look almost the same, can you use one inside the other?